cycler/__init__.py

```diff
@@ -180,21 +201,27 @@ def change_key(self, old, new):
             return
         if new in self._keys:
             raise ValueError(
-                "Can't replace {old} with {new}, {new} is already a key"
-                .format(old=old, new=new)
+                f"Can't replace {old} with {new}, {new} is already a key"
             )
         if old not in self._keys:
-            raise KeyError("Can't replace {old} with {new}, {old} is not a key"
-                           .format(old=old, new=new))
+            raise KeyError(
+                f"Can't replace {old} with {new}, {old} is not a key"
+            )
 
         self._keys.remove(old)
         self._keys.add(new)
 
-        if self._right is not None and old in self._right.keys:
+        if (
+            self._right is not None
+            and isinstance(self._right, Cycler)
```

Comment:
explicitly checking isinstance was required for mypy to pass as the option of `list[dict[K, V]]` which is a seemingly valid type (at least tests blew up when I removed the lines from init which added that type for `_right`)

That said, if it is valid for `_right` to be this flattened list of dict instead of a full cycler object, should probably handle the rename like we do for `_left`

And if it is invalid, should probably figure out why the code is there for it in `__init__`

Reply:
Upon more careful inspection, the only test which failed was the one that was called `test_starange_init` [sic] which is not the recommended way of initializing a cycler, and the cycler object created in that test fails when you try to rename the key of the right child.

So we can:
- explicitly disallow right from being an iterable of dicts (while left can be for various reasons)
  - this would simplify this and some other checks as the only options are cycler or None
  - change the test to reflect that this is disallowed
- explicitly allow this formation, and add tests and correct code where _right is accessed, including at least `change_key`
  - add to this test or the change key test a test for this case

Reply:
Sidenote: this is actually a pretty reasonable example of the kind of problem that `mypy` can make much more obvious and are not caught, even with "100% of lines coverage" as we have here.

Failing code (adding one line to the strange init test, run on main, this PR would actually make it fall through):

```python
>>> from cycler import cycler, Cycler
>>> c = cycler('r', 'rgb')
>>> c2 = cycler('lw', range(3))
>>> cy = Cycler(list(c), list(c2), zip)
>>> cy.change_key("lw", "ls")
╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ <stdin>:1 in <module>                                                                            │
│                                                                                                  │
│ /home/kyle/src/scipy/cycler/cycler.py:193 in change_key                                          │
│                                                                                                  │
│   190 │   │   self._keys.remove(old)                                                             │
│   191 │   │   self._keys.add(new)                                                                │
│   192 │   │                                                                                      │
│ ❱ 193 │   │   if self._right is not None and old in self._right.keys:                            │
│   194 │   │   │   self._right.change_key(old, new)                                               │
│   195 │   │                                                                                      │
│   196 │   │   # self._left should always be non-None                                             │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
AttributeError: 'list' object has no attribute 'keys'
```

Reply:
I actually do think this is still unresolved... it passes tests, but the above test case would have the unexpected result of not actually changing the key.

Can `right` be disallowed from being `iterable[dict]`? (therefore changing that test)
Or, should the case of `right` being `iterable[dict]` be handled as it is for `left`?